### PR TITLE
XC Multihop proposal document transfer details and improvements 

### PR DESCRIPTION
**Pull request by dzmitry-lahoda**
- [x] PR title is my best effort to provide summary of changes and has clear text to be part of release notes 
- [x] I marked PR by `misc` label if it should not be in release notes
- [x] I have linked Zenhub/Github or any other reference item if one exists
- [x] I was clear on what type of deployment required to release my changes (node, runtime, contract, indexer, on chain operation, frontend, infrastructure) if any in PR title or description
- [x] I waited and did best effort for `pr-workflow-check / draft-release-check` to finish with success(green check mark) with my changes
- [x] I have added at least one reviewer in reviewers list
- [x] I tagged(@) or used other form of notification of one person who I think can handle best review of this PR
- [x] I have proved that PR has no general regressions of relevant features and processes required to release into production

### CODE REVIEW

File: rfcs/0015-one-signature-multihop-transfer-solution-ibc-xcm.md [hunk 1]
```diff
@@ -150,11 +149,21 @@ Cannot express(securely) arbitrary routing and programs execution.
 
 To be handled in next RFC or implementation.
 
+## Error handling
```
Comment (blasrodri) [inline rfcs/0015-one-signature-multihop-transfer-solution-ibc-xcm.md:152]: Can you please provide an example where the unwinding happens, and it succeeds? And one where it doesn't?

**Comment by blasrodri** [comment]
Please let's add a few examples, with some graphics (ascii art is also welcome)

**Comment by RustNinja** [comment]
@dzmitry-lahoda @blasrodri @josietyleung 
Me as a Polkadot account who wants to initiate transfers from DOT to Osmosis/Banksy.
- i will create the xcm call reserve transfer assets and will specify multi-location.
- Without rollbackk/bad path when ibc will fail when deliver packet from composable to picasso then my token blocked on composable. I do not have any LAYR token to initiate tx and move DOT from composable back to polkadot or via IBC further to picasso via IBC.

current version of pallet ibc does not have ibc message forwarding with memo and does

**Comment by blasrodri** [comment]
Seems like we need to implement https://github.com/strangelove-ventures/packet-forward-middleware but for both pallet-ibc and the relationship between ibc + xcm

File: rfcs/0015-one-signature-multihop-transfer-solution-ibc-xcm.md (hunk 1 above)
Comment (dzmitry-lahoda) [reply rfcs/0015-one-signature-multihop-transfer-solution-ibc-xcm.md:152]: sure, easy. just have read reference module. see ok. will unwind as needed if to do in Substrate same as in Go. Plus final original tracking of sent xcm packet (basically unwind for xcm in hook).

**Comment by dzmitry-lahoda** [comment]
> current version of pallet ibc does not have ibc message forwarding with memo and does

exactly, this should be implemted, ported from referenced go code

**Comment by dzmitry-lahoda** [comment]
> Seems like we need to implement https://github.com/strangelove-ventures/packet-forward-middleware but for both pallet-ibc and the relationship between ibc + xcm

@blasrodri yes. that was idea. port some Go code to XCM.

**Review by RustNinja — APPROVED** [review]

**Comment by dzmitry-lahoda** [comment]
@RustNinja may you summarize all quesitions remaining here so we align these in this PR?

```
Nikita  [3:43 PM]
I have a few question about Batch tx.
Who are gonna change the banksy to make a batch tx?
Does the batch tx should work for any chain?
Or just for our set of chains?
polkadot - composable - picasso - banksy - osmo? and reverse
Is it for any tokens? or is just for dot? all tokens
why do we need
```
What else? (on top of request how we handle XCM rollbacks

**Comment by RustNinja** [comment]
> 



> @RustNinja may you summarize all quesitions remaining here so we align these in this PR?
> 
> ```
> Nikita  [3:43 PM]
> I have a few question about Batch tx.
> Who are gonna change the banksy to make a batch tx?
> Does the batch tx should work for any chain?
> Or just for our set of chains?
> polkadot - composable - picasso - banksy - osmo? and reverse
> Is it for any tokens? or is just for dot? all tokens
> why do we need
> ```
> 
> What else? (on top of request how we handle XCM rollbacks

thanks. they all are answered already.

yes. xcm rollbacks is very useful. Thank you so much.

**Comment by dzmitry-lahoda** [comment]
Added some, will consistent little bit. Not sure about amount of details for making all IBC middleware XCM. We far behind official ibc-rs. Seems will need some glue to hook some callbacks (middleware actually) into pallet-ibc. I do not specify that. It is pure grunt eng work :)

**Review by RustNinja — APPROVED** [review]

**Comment by dzmitry-lahoda** [comment]
https://github.com/strangelove-ventures/packet-forward-middleware/pull/87

**Comment by dzmitry-lahoda** [comment]
no reaction, nice

**Review by blasrodri — APPROVED** [review]

**Review by RustNinja — APPROVED** [review]

**Comment by dzmitry-lahoda** [comment]
execute can be only called by root, so cannot do arbitrary txes yet.

**Comment by RustNinja** [comment]
> execute can be only called by root, so cannot do arbitrary txes yet.
Thanks.
Does it mean that we need to use `reserve_transfer_assets` with a custom multilocation (pallets and index) to move assets from polkadot/substrate chain?

**Comment by RustNinja** [comment]
@dzmitry-lahoda 
another question is how to initiate the ibc/xcm call? 

there are 2 options:

- should it be in the same block when ibc/xcm instruction was executed? 
as a callback function in the same block?

- Does pallet xcm ibc pallet need just update the storage in callback and then 
next block on_finalize hook will trigger new xcm/ibc packet?

which option do you prefer?

**Comment by dzmitry-lahoda** [comment]
in same block immediately. that is simpler. llwe did not do xcm benches, but can run and make bench depend on account to incroporate ibc. or can optimize like you said.

**Comment by dzmitry-lahoda** [comment]
where to encode index staff, dest or beneficiary, what works.

**Comment by RustNinja** [comment]
@dzmitry-lahoda 
https://github.com/ComposableFi/composable/blob/main/rfcs/0015-one-signature-multihop-transfer-solution-ibc-xcm.md#centauricosmos-sdk------composable-picassosubstrate---composable-composablesubstrate---bifrostsubstrate

i want to confirm:
```
{
  "forward": {
    "receiver": "composable-alice",
    "port": "transfer",
    "channel": "channel-123",
    "next": {
      "substrate" : {
          "parent" : "Polkadot",
          "parachain" : "Bifrost",
          "account" : "alice", 
      }
    }
  }
}
```

am i right that substrate hop can be only the last hop in memo?
Because other substrate chains do not support multi hop so in our example composable or picasso will able to forward message once directly to polkadot or kusama or bifrost... 
that's all

**Pull request merged**